lib/hanami/view/html.rb

```diff
@@ -0,0 +1,104 @@
+# frozen_string_literal: true
+
+module Hanami
+  class View
+    module HTML
+      # A string that has been marked as "HTML safe", ensuring that it is not automatically escaped
+      # when used in HTML view templates.
+      #
+      # A SafeString is frozen when initialized to ensure it cannot be mutated after being marked as
+      # safe, which could be an avenue for injection of unsafe content.
+      #
+      # @see String#html_safe
+      #
+      # @api public
+      # @since 2.0.0
+      class SafeString < String
+        # @api public
+        # @since 2.0.0
+        def initialize(string)
+          super(string)
+          freeze
+        end
+
+        # @api private
+        private def initialize_copy(other)
+          super
+          freeze
+        end
+
+        # @return [true]
+        #
+        # @api public
+        # @since 2.0.0
+        def html_safe?
+          true
+        end
+
+        # @return [self]
+        #
+        # @api public
+        # @since 2.0.0
+        def html_safe
+          self
+        end
+
+        # @return [self]
+        #
+        # @api public
+        # @since 2.0.0
+        def to_s
+          self
+        end
+      end
+
+      # @api private
+      # @since 2.0.0
+      module StringExtensions
+        # Returns the string as a {Hanami::View::HTML::SafeString}, ensuring the string is not
+        # automatically escaped when used in HTML view templates.
+        #
+        # @return [Hanami::View::HTML::SafeString]
+        #
+        # @api public
+        # @since 2.0.0
+        def html_safe
+          Hanami::View::HTML::SafeString.new(self)
+        end
+      end
+    end
+  end
+end
+
+class String
```

Comment:
@timriley Do we have a chance not to do this?

In the context where `#html_safe?` is invoked, can we hook it into a wrapper?

Naive code example:

```ruby
class Wrapper
  def call(obj)
    case obj
    when String
        # ...
    when Numeric
       # ....
    end
  end
end

# This is a fictional object from Temple/Tilt where the escaping happens
class TempleRenderingContext
  def call(obj)
     wrapper.call(obj)
  end
end
```

Reply:
@jodosha Fair question, thank you for asking!

I'm not sure if we have much choice here, because we don't really control every instance of “the context where `#html_safe?` is invoked” — for the three template engines we support (Hanami::View::ERB, Slim, Haml) this chiefly happens [inside Temple's Escapable filter](https://github.com/judofyr/temple/blob/f8a069bdba5d2d821b1a8c8221b9e6e225e9e1ed/lib/temple/filters/escapable.rb#L18).

For Hanami::View::ERB, we swap this filter for one that we fully own, and which uses an approach like you describe, but that still wouldn't help us in terms of providing seamless support for Haml and Slim. They both use the Temple Escapable filter directly, and Haml also has [an additional check](https://github.com/haml/haml/blob/d0a86c3eba4563073e950f9e7307a947d4379a64/lib/haml/util.rb#L31) in its escaping utils. To change this would require us to maintain patches of Haml and Slim that are delicate and prone to breakage.

I'm sure it'd be possible for us to come up with a whole new strategy and slowly introduce all the required pieces into Temple, Haml, Slim and the like, but that feels like a lot of time-consuming work simply for the sake of avoiding this one String/Object/Numeric extension that is already endemic across the Ruby web app ecosystem.

In addition, your suggestion here doesn't cover the case where user-authored helpers might need to return strings flagged as html_safe. In this case, if we decide not to use the `String#html_safe` extension, then users would need to construct their own SafeString instances directly. We could probably provide a helper available in _most_ contexts to do this, e.g. `html_safe(my_str)`, but it would only cover those helpers working in execution contexts that the Hanami framework expressly controls. In other contexts, our users would have to remember to `Hanami::View::HTML::SafeString.new(my_str)` instead. This approach doesn't feel great to me.

So my opinion here is that, on balance, this is one place where we're better off simply following the already-established and widespread convention. Folks are going to be learning a whole bunch of stuff already when they adopt Hanami. Here we can at least take one thing off that list :)

Reply:
I'm going to go ahead and merge this PR without any changes here, but I do want to make it clear that I'm definitely open to adopting a different approach if it match the usability and robustness of the current one.

(I won't be pursuing that personally because, as I noted above, I'm pretty happy with the current arrangement, and there's a bunch of other much more pressing things to get done to complete our view layer for 2.0)

Reply:
@timriley Thanks for the extensive explanation. At this point, it’s the most pragmatic path to take.

We’ll reuse your words in our guides as a note for this monkey-patching exception.